test/test_strerror.cpp

```diff
@@ -12,6 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#ifdef _GNU_SOURCE
+#undef _GNU_SOURCE
+#include <string>
+#else
+#include <string>
+#endif
```

Comment:
@brawner @wjwwood @clalancette this is a non-portable workaround for an issue @Blast545 and I stumbled upon. Turns out `librcutls.so` and this test's executable where linking against different `strerror_r` versions. Different `libc` versions, and different symbol names (`strerror_r` is actually a macro for the `libc` version `librcutils.so` links to on Ubuntu 20.04).

This seems like a significant pain point whenever mocking system libraries if it turns out such practices are widespread (and I'd think so). We need a way to ensure both test and library use and link against the same third party OR use a wrapper with a stable ABI instead of system libraries directly.

Reply:
> @brawner @wjwwood @clalancette this is a non-portable workaround for an issue @Blast545 and I stumbled upon. Turns out `librcutls.so` and this test's executable where linking against different `strerror_r` versions.

Huh, weird.  Can you explain a little more of the details?  I'm just trying to get a handle on how common this will be.

Reply:
_holiday mode OFF_ 

> Can you explain a little more of the details? I'm just trying to get a handle on how common this will be.

Maybe ELF dumps help.

Output fragment of `readelf --relocs path/to/librcutils.so`:
```
Relocation section '.rela.plt' at offset 0x3480 contains 107 entries:
  Offset          Info           Type           Sym. Value    Sym. Name + Addend
...
0000000180c8  000e00000007 R_X86_64_JUMP_SLO 0000000000000000 __xpg_strerror_r@GLIBC_2.3.4 + 0
000000018358  007000000007 R_X86_64_JUMP_SLO 000000000000e250 rcutils_strerror + 0
...
```

Output fragment of `readelf --relocs path/to/test_strerror`:
```
Relocation section '.rela.plt' at offset 0x6780 contains 206 entries:
  Offset          Info           Type           Sym. Value    Sym. Name + Addend
...
0000000799e8  001700000007 R_X86_64_JUMP_SLO 0000000000000000 strerror@GLIBC_2.2.5 + 0
000000079bd8  005700000007 R_X86_64_JUMP_SLO 0000000000000000 rcutils_strerror + 0
...
```
Relevant `glibc` [`string.h`](https://github.molgen.mpg.de/git-mirror/glibc/blob/20003c49884422da7ffbc459cdeee768a6fee07b/string/string.h#L406-L441) header section.

I don't know exactly why this happens, though the C++ runtime may have something do with it (`librcutils.so` is C only).

_holiday mode on_

Reply:
@Blast545 we need a solution for this. Playing with system library defines is brittle and non-portable. 

Digging a bit, it turns out that it's `libstdc++` that **requires** `_GNU_SOURCE` to be set (see [here](https://gcc.gnu.org/onlinedocs/libstdc++/faq.html#faq.predefined)). It's admittedly a bad workaround. We could set `_GNU_SOURCE` when building `rcutils` on Linux. @clalancette @dirk-thomas @wjwwood for  feedback.

Reply:
After talking to @hidmic about this offline, I offered this opinion:

I think so long as a) we do not require these GNU specific functions (graceful degradation to POSIX or Windows is possible) and b) they (the GNU functions) are available, then it is fine to use them or force it to use them.

I don't have a sense for any intrinsic drawback or advantage to using the GNU ones over the POSIX ones.
Like performance or security benefits, but they could convince me otherwise, if they exist.

Reply:
PTAL https://github.com/ros2/rcutils/pull/267. Enabling GNU extensions has two potential drawbacks: (1) it makes it possible for non-portable C APIs to be used, and (2) it seems GNU folks departed from the POSIX standard for some functions. We can consciously avoid (1), but I have yet to find an exhaustive list of such departures.

Reply:
Ultimately, the problem's in `g++`, but the bug's been standing for 18 years and is tagged as `wontfix`. It also creates some fun booby traps in C++ code I wasn't aware off e.g. you can't use the stl and name a free function `minor` headers, it'll clash with some GNU macro.

Reply:
For the most part, I agree with @wjwwood 's assessment.  My only concern is about booby-traps like @hidmic laid out in the last comment.

There are specific areas I'm worried about:
1.  The cross-platform area.  If you use a non-portable GNU API by accident, will it obviously show up (fail to compile or link) on the other platforms?
1.  The name clash areas.  As @hidmic found out, it looks like the GNU API is "wider" than the POSIX one.  Is it obvious when you run into one of these clashes?  Have we tried running a full CI run to see what happens?

Assuming we are happy with the answers to those questions, then yeah, I'm fine with going ahead with the GNU API on Linux.

Reply:
> The cross-platform area. If you use a non-portable GNU API by accident, will it obviously show up (fail to compile or link) on the other platforms?

In general, yes. But there are [other cases](https://www.gnu.org/prep/standards/html_node/Non_002dGNU-Standards.html), more subtle, that we have to bear in mind.

> The name clash areas. As @hidmic found out, it looks like the GNU API is "wider" than the POSIX one. Is it obvious when you run into one of these clashes? Have we tried running a full CI run to see what happens?

So far, `rcutils` builds and tests succeed locally. That does not guarantee we won't ever come across a name clash (with a macro, other redefinitions will be caught by the compiler). 

---

Nevermind the choice we make here, I'd want to stress the fact that, because of what `g++` does when it finds `glibc` in the system, in general we cannot ignore GNU API peculiarities. A C++ program using `getopt` is potentially broken already if these are not taken into account.

Reply:
> Nevermind the choice we make here, I'd want to stress the fact that, because of what `g++` does when it finds `glibc` in the system, in general we cannot ignore GNU API peculiarities. A C++ program using `getopt` is potentially broken already if these are not taken into account.

Actually, that brings up another point to consider here.  I don't remember the exact way to do it in cmake/ament, but I think we need to "export" `-D_GNU_SOURCE` to upstream packages so that they also compile properly.  Does that sound right to you?

Reply:
That's probably the safest bet, but I fear we're spreading the disease. Should a downstream package that uses `rcl` be exposed to GNU extensions?